assets/shaders/RayTracing.rgen

```diff
@@ -25,79 +25,74 @@ void main()
-		const vec2 pixel = vec2(gl_LaunchIDEXT.x + RandomFloat(pixelRandomSeed), gl_LaunchIDEXT.y + RandomFloat(pixelRandomSeed));
-		const vec2 uv = (pixel / gl_LaunchSizeEXT.xy) * 2.0 - 1.0;
-
-		vec2 offset = Camera.Aperture/2 * RandomInUnitDisk(Ray.RandomSeed);
-		vec4 origin = Camera.ModelViewInverse * vec4(offset, 0, 1);
-		vec4 target = Camera.ProjectionInverse * (vec4(uv.x, uv.y, 1, 1));
-		vec4 direction = Camera.ModelViewInverse * vec4(normalize(target.xyz * Camera.FocusDistance - vec3(offset, 0)), 0);
-		vec3 rayColor = vec3(1);
-
-		// Ray scatters are handled in this loop. There are no recursive traceRayEXT() calls in other shaders.
-		for (uint b = 0; b <= Camera.NumberOfBounces; ++b)
+		const float tMin = 0.001;
+		const float tMax = 10000.0;
+
+		// If we've exceeded the ray bounce limit without hitting a light source, no light is gathered.
+		// Light emitting materials never scatter in this implementation, allowing us to make this logical shortcut.
+		if (b == Camera.NumberOfBounces) 
 		{
-			const float tMin = 0.001;
-			const float tMax = 10000.0;
-
-			// If we've exceeded the ray bounce limit without hitting a light source, no light is gathered.
-			// Light emitting materials never scatter in this implementation, allowing us to make this logical shortcut.
-			if (b == Camera.NumberOfBounces) 
-			{
-				rayColor = vec3(0, 0, 0);
-				break;
-			}
-
-			traceRayEXT(
-				Scene, gl_RayFlagsOpaqueEXT, 0xff, 
-				0 /*sbtRecordOffset*/, 0 /*sbtRecordStride*/, 0 /*missIndex*/, 
-				origin.xyz, tMin, direction.xyz, tMax, 0 /*payload*/);
-			
-			const vec3 hitColor = Ray.ColorAndDistance.rgb;
-			const float t = Ray.ColorAndDistance.w;
-			const bool isScattered = Ray.ScatterDirection.w > 0;
-
-			rayColor *= hitColor;
-
-			// Trace missed, or end of trace.
-			if (t < 0 || !isScattered)
-			{				
-				break;
-			}
-
-			// Trace hit.
-			origin = origin + t * direction;
-			direction = vec4(Ray.ScatterDirection.xyz, 0);
+			rayColor = vec3(0, 0, 0);
+			break;
 		}
 
-		pixelColor += rayColor;
-	}
+		traceRayEXT(
+			Scene, gl_RayFlagsOpaqueEXT, 0xff, 
+			0 /*sbtRecordOffset*/, 0 /*sbtRecordStride*/, 0 /*missIndex*/, 
+			origin.xyz, tMin, direction.xyz, tMax, 0 /*payload*/);
+		
+		const vec3 hitColor = Ray.ColorAndDistance.rgb;
+		const float t = Ray.ColorAndDistance.w;
+		const bool isScattered = Ray.ScatterDirection.w > 0;
+
+		rayColor *= hitColor;
+
+		// Trace missed, or end of trace.
+		if (t < 0 || !isScattered)
+		{				
+			break;
+		}
+
+		// Trace hit.
+		origin = origin + t * direction;
+		direction = vec4(Ray.ScatterDirection.xyz, 0);
+	} 
 
-	const bool accumulate = Camera.NumberOfSamples != Camera.TotalNumberOfSamples;
-	const vec3 accumulatedColor = (accumulate ? imageLoad(AccumulationImage, ivec2(gl_LaunchIDEXT.xy)) : vec4(0)).rgb + pixelColor;
+	//Only reset the image on the first thread touching it.
+	const bool accumulate = Camera.NumberOfSamples != Camera.TotalNumberOfSamples
```

Comment:
If it's anything like in CUDA, this is likely a race condition as you have no guarantee that the `z == 0` threads are run before the `z != 0` threads. So you are probably losing samples on the first image (when resetting the accumulation buffer).

Unless the `vkCmdTraceRaysKHR` called in a loop with the incremented depth is accounting for that? Not sure if it requires any memory barrier then.

Reply:
Yes, this is indeed a race condition, but the only consequence is potentially loosing a few random samples on the debug view and on the first frame after reset.

There might be a way to do this properly by introducing an VK_FORMAT_R32_UINT texture to collect the samples per pixel, but I'm unsure if it's worth it if this race condition isn't triggered in practice.

Reply:
I just remembered, that I still should have code for that around, I will try to implement it.